### PR TITLE
feature: auto updater!

### PR DESCRIPTION
**Pull request by diamante0018**
**What does this PR do?**

- Auto updater
- Fixes singleton component memory leak

### CODE REVIEW

File: src/Components/Modules/Updater.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		UIScript::Add("getAutoUpdate", [](const UIScript::Token& /*token*/, const Game::uiInfo_s* /*info*/)
+		{
+			if (const auto exe = GetLauncher(); exe.has_value())
+			{
+				Game::Sys_QuitAndStartProcess(exe.value().data());
+				return;
+			}
+
+			// No launcher was found on the system, time to tell them to download it from GitHub
+			Utils::OpenUrl("https://forum.alterware.dev/t/how-to-install-the-alterware-launcher/56");
```
Comment (Rackover) [inline src/Components/Modules/Updater.cpp:103]: Sorry to be extra annoying but i would like to ask you to please turn all those addresses and filename (http://forum, alterware-launcher.exe, the github repo URL, etc...) into variables or macros at the beginning of the file or in the headerfile, however you like it. thanks! 

Other than that it's good to go, so whenever it's done i will test it and merge it.
Does this have a dependency on any IW4x rawfile PR you would want this to get tested with? So I can test both together & greenlight both!

File: src/Components/Modules/Updater.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		// If they use the alterware-launcher once to install they will have this file
+		// If they don't, what are they waiting for?
+		constexpr auto* revisionFile = ".iw4xrevision";
+
+		void CheckForUpdate()
+		{
+			std::string revision;
+			if (!Utils::IO::ReadFile(revisionFile, &revision) || revision.empty())
+			{
+				Logger::Print(".iw4xrevision does not exist. Notifying the user an update is available\n");
```
Comment (Rackover) [inline src/Components/Modules/Updater.cpp:27]: missed an occasion to std::format here
`Logger::Print("{} does not exist", revisionFile)`

**Review by Rackover — CHANGES_REQUESTED** [review]

File: src/Components/Modules/Updater.cpp (hunk 1 above)
Comment (diamante0018) [reply src/Components/Modules/Updater.cpp]: Yeah both should be merged at the same time

File: src/Components/Modules/Updater.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+		// If they use the alterware-launcher once to install they will have this file
+		// If they don't, what are they waiting for?
+		constexpr auto* REVISION_FILE = ".iw4xrevision";
+		constexpr auto* GITHUB_REMOTE_URL = "https://api.github.com/repos/iw4x/iw4x-client/releases/latest";
+		constexpr auto* INSTALL_GUIDE_REMOTE_URL = "https://forum.alterware.dev/t/how-to-install-the-alterware-launcher/56";
+
+		void CheckForUpdate()
+		{
+			std::string revision;
```
Comment (Rackover) [inline src/Components/Modules/Updater.cpp:24]: no `{}` ?

File: src/Components/Modules/Updater.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			{
+				// Nothing to do in this situation. We won't know if we need to update or not
+				Logger::Print("GitHub sent an invalid reply\n");
+				return;
+			}
+
+			if (!doc.HasMember("tag_name") || !doc["tag_name"].IsString())
+			{
+				// Nothing to do in this situation. We won't know if we need to update or not
+				Logger::Print("GitHub sent an invalid reply\n");
```
Comment (Rackover) [inline src/Components/Modules/Updater.cpp:51]: Its preferable not to have two different errors print out the same message. You can add (malformed JSON) on the first one and (missing tag_name) on the second for instance.

**Review by Rackover — CHANGES_REQUESTED** [review]
Two small things to address and then i'll merge it with my eyes closed

File: src/Components/Modules/Updater.cpp (hunk 3 above)
Comment (diamante0018) [reply src/Components/Modules/Updater.cpp:24]: That has no functional change: {} -> calls the default ctor which is the same as my code.

**Review by Rackover — APPROVED** [review]

**Pull request merged**